File: contracts/genlayer/MarketResolver.py
# { "Depends": "py-genlayer:1jb45aa8ynh2a9c9xn3b7qqh8sm5q93hwfp7jqmwsfhh8jpz09h6" }

from genlayer import *

import json


class MarketResolver(gl.Contract):
    def __init__(self):
        pass

    @gl.public.write
    def resolve(
        self,
        question: str,
        resolution_criteria: str,
        resolution_source_url: str,
        resolves_at: str,
    ) -> str:
        if not resolution_source_url:
            return json.dumps(
                {
                    "outcome": "UNKNOWN",
                    "confidence": 0,
                    "evidence": "No resolution source URL was provided.",
                    "sourceUsed": "",
                    "reasoning": "A named source is required for GenLayer resolution.",
                },
                sort_keys=True,
            )

        def nondet() -> str:
            response = gl.nondet.web.get(resolution_source_url)
            source_body = response.body.decode("utf-8")
            prompt = f"""
Resolve this prediction market using only the named source.

Question: {question}
Resolution criteria: {resolution_criteria}
Resolution source URL: {resolution_source_url}
Scheduled resolution time: {resolves_at}

Named source page content:
{source_body}

Rules:
- Resolve YES only if the named source clearly confirms the criteria.
- Resolve NO only if the named source clearly contradicts the criteria.
- Return UNKNOWN if the source is unreachable, blocked, ambiguous, paywalled, login-only, stale, or insufficient.
- Never guess.

Respond only with valid JSON matching this shape:
{{
  "outcome": "YES" | "NO" | "UNKNOWN",
  "confidence": number,
  "evidence": string,
  "sourceUsed": string,
  "reasoning": string
}}
"""
            result = gl.nondet.exec_prompt(prompt).replace("```json", "").replace("```", "").strip()
            return json.dumps(json.loads(result), sort_keys=True)

        return gl.eq_principle.strict_eq(nondet)

File: src/lib/genlayer/marketResolver.ts
import { z } from "zod";
import { callGenLayer, type GenLayerResult } from "./client";

export const genLayerMarketResolverInputSchema = z.object({
  question: z.string().min(1),
  resolutionCriteria: z.string().min(1),
  resolutionSourceUrl: z.string().url(),
  resolvesAt: z.string().min(1),
});

export const genLayerResolutionSchema = z.object({
  outcome: z.enum(["YES", "NO", "UNKNOWN"]),
  confidence: z.number().min(0).max(100),
  evidence: z.string(),
  sourceUsed: z.string(),
  reasoning: z.string(),
});

export type GenLayerMarketResolverInput = z.infer<typeof genLayerMarketResolverInputSchema>;
export type GenLayerResolution = z.infer<typeof genLayerResolutionSchema>;

export async function requestGenLayerResolution(input: GenLayerMarketResolverInput): Promise<GenLayerResult<GenLayerResolution>> {
  const parsed = genLayerMarketResolverInputSchema.safeParse(input);
  if (!parsed.success) {
    return { status: "ERROR", error: parsed.error.message };
  }

  return callGenLayer("MARKET_RESOLVER", "genlayer_requestMarketResolution", {
    ...parsed.data,
    rules: [
      "Resolve YES only if the named source clearly confirms the criteria.",
      "Resolve NO only if the named source clearly contradicts the criteria.",
      "Return UNKNOWN if the source is unreachable, blocked, ambiguous, paywalled, login-only, stale, or insufficient.",
      "Never guess.",
    ],
  }, validateGenLayerResolution);
}

export async function checkGenLayerResolution(txHash: string): Promise<GenLayerResult<GenLayerResolution>> {
  if (!txHash) return { status: "ERROR", error: "txHash is required" };
  return callGenLayer("MARKET_RESOLVER", "genlayer_checkMarketResolution", { txHash }, validateGenLayerResolution);
}

export function validateGenLayerResolution(value: unknown): GenLayerResolution {
  return genLayerResolutionSchema.parse(value);
}
